plugins/org.jboss.tools.openshift.cdk.server/src/org/jboss/tools/openshift/internal/crc/server/ui/PullSecretValidation.java

```diff
@@ -0,0 +1,126 @@
+import java.io.UncheckedIOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.osgi.util.NLS;
+import org.jboss.tools.openshift.internal.common.ui.job.UIUpdatingJob;
+import org.json.JSONArray;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+public class PullSecretValidation {
+
+	private static final int MAX_PULLSECRET_FILE_SIZE = 20 * 1024; // 20k
+
+	private final AtomicReference<String> error = new AtomicReference<>();
+	private final AtomicReference<String> file = new AtomicReference<>();
+
+	public String validate(final String file, final Consumer<String> listener) {
+		if (!hasChanged(file)) {
+			return error.get();
+		}
+		error.set(null);
+		this.file.set(file); 
+		new PullSecretValidationJob(file, listener).schedule();
+		return error.get();
+	}
+
+	private boolean hasChanged(String file) {
+		return !Objects.equals(this.file.get(), file);
+	}
+
+	private class PullSecretValidationJob extends UIUpdatingJob {
+
+		private String pullSecretFile;
+		private Consumer<String> listener;
+		
+		private PullSecretValidationJob(String pullSecretFile, Consumer<String> listener) {
+			super("Validating pull secret...");
+			this.pullSecretFile = pullSecretFile;
+			this.listener = listener;
+		}
+
+		@Override
+		protected IStatus run(IProgressMonitor monitor) {
+			String errorMessage = validate(pullSecretFile);
+			error.set(errorMessage);
+			return Status.OK_STATUS;
+		}
+
+		@Override
+		protected IStatus updateUI(IProgressMonitor monitor) {
+			listener.accept(error.get());
+			return Status.OK_STATUS;
+		}
+
+		private String validate(String file) {
+			String msg = null;
+			if (file == null || file.isEmpty() || !(new File(file)).isFile()) {
+				msg = "Please select a valid Pull Secret file. See https://cloud.redhat.com/openshift/install/crc/installer-provisioned for instructions.";
+			} else if (!new File(file).canRead()) {
+				msg = "Pull Secret file is not readable.";
+			} else {
+				msg = validateJsob(file);
+			}
+			return msg;
+		}
+
+		private String validateJsob(String file) {
+			String msg = null;
+			try {
+				if (!isValidJson(readPullSecretFile(file))) {
+					msg = NLS.bind("Pull Secret file {0} is invalid. Make sure it is valid JSON.", file);
+				}
+			} catch (IOException | UncheckedIOException e) {
+				msg = NLS.bind("Pull Secret file {0} is invalid. Make sure it is valid JSON: {1}", file, e.getLocalizedMessage());
+			}
+			return msg;
+		}
+
+		private String readPullSecretFile(String pullSecretFile) throws IOException {
+			String content = null;
+			try (Stream<String> lines = Files.lines(Paths.get(pullSecretFile))) {
```

Comment:
I am getting proper validation I have to say. When I try to pass big binary file I get right message (Pull secret file /b/b/b is invalid. Make sure it is valud JSON: ) and then I get error message that was caught in try catch. java.nio.charset.MalformedInputException: Input length = 1.

This would mean that the file read failed due to charset exception (aa binary is not standard UTF-8 coded file, which is ok) but I think that this is not proper validation. Like maybe it is intentional, not sure here. I quite do not understand the limit then since the real validtion is based on if it is text file or binary (thus throwing charset exception).

Reply:
the problem here was that the code tried to load the whole file and then caused an out-of-error. I am not validating this, I'm just preventing this by assuming that a pull secret file would not be larger than 20k. We could test the file size beforehand and refuse to load anything that's larger than 20k and would be then explicit about this limitation.
The validation error that you see is caused by reading a binary file where a text file is expected. I could catch this error and use a more userfriendly error "not a text file" or something.

Reply:
I tried latest change -> It filters files < 20 K bytes, which is fine now. I would still add a catch for malformedInputException which, as you said, is exactly reading a binary (not text file), and show more user friendly message. Thanks.

Reply:
@odockal : I'm now catching this error and displaying a message that says that the file isn't a text file. Please review